### DISCUSSION TITLE
ggplot2 extension advent of code: set_theme_last_plot()

### DISCUSSION THREAD

**Discussion by EvaMaeRey** [ggplot2-extenders/ggplot-extension-club#67] [General]
Thinking about this workflow, I thought his might be a good 'advent of code' challenge:  https://bsky.app/profile/gforb.bsky.social/post/3lciojslcuc2t 

'Copy and pasting the same + theme(...) line into every plot.'

I'd though about a variant: 1) build plot (explore data) 1.b) make plot look good; adjust plot w/ theme() 2) set theme based on last plot w/ set_theme_last_plot() 3) keep plotting.  

I didn't manage to write a working set_theme_last_plot()!  I think I need to make `t1` look like `t2` before trying to write the set_theme_last_plot() function.

``` r
library(ggplot2)
ggplot(data = cars) +
  aes(x = speed, y = dist) + 
  geom_point() + 
  theme(panel.grid = ggplot2::element_blank())
```

![](https://i.imgur.com/XtMkz4t.png)

``` r

t1 <- last_plot()$theme  # add class theme and gg

t2 <- ggplot2::theme(
        panel.grid = ggplot2::element_blank(),
        complete = TRUE 
      ) 

waldo::compare(t1, t2)
#> `old` is a list
#> `new` is an S3 object of class <theme/gg>, a list
```

<sup>Created on 2024-12-09 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>

**Comment by yjunechoe** [comment]
Fun experiment! I don't have anything to contribute because I rarely use `theme()`, but can I suggest some names? :D

- `persist_theme()`
- `recycle_theme()`
- `repeat_theme()`

Or perhaps less side-effect-y:

```r
plot1 + theme_linedraw()

plot2 + theme_last_plot()
```

**Reply by EvaMaeRey** [reply]
Nice ideas!

**Comment by teunbrand** [comment]
Just here to point that the obvious solution doesn't work.
The reason is that `ggplot()` uses `set_last_plot()` internally, so when you're requesting the theme of the previous plot, while building the next plot, it is going to point to itself. I.e. in the example below, `last_theme()` retrieves the theme in `ggplot(mtcars, ...)`.

``` r
library(ggplot2)

last_theme <- function(plot = last_plot()) plot$theme

ggplot(mpg, aes(displ, hwy)) +
  geom_point() +
  theme(text = element_text(colour = "red"))
```

![](https://i.imgur.com/KUnwZJZ.png)

``` r
ggplot(mtcars, aes(disp, mpg)) +
  geom_point() +
  last_theme() +
  theme(line = element_line(linetype = "dotted"))
```

![](https://i.imgur.com/qWgYOcb.png)

<sup>Created on 2024-12-09 with [reprex v2.1.1](https://reprex.tidyverse.org)</sup>

**Reply by yjunechoe** [reply]
Wow and all this time I thought `last_plot()` was set on rendering/`ggplot_build()`, not on evaluating `ggplot()`!

**Reply by yjunechoe** [reply]
Actually it seems to apply to both cases? But setting last *rendered* plot seems to behave slightly differently, in not recovering the last ggplot object exactly as-is? (is there an actual difference or is this an artifact?)

```r
library(ggplot2)
p1 <- ggplot() + labs(title = "p1")
p2 <- ggplot() + labs(title = "p2")

# Case 1: `last_plot()` retrieves last ggplot object created by `ggplot()`
waldo::compare(last_plot(), p2)
#> ✔ No differences

# Case 2: `last_plot()` retrieves last ggplot rendered via `print()`/`plot()`
# - But not strictly equivalent
plot(p1)
waldo::compare(last_plot(), p1, ignore_attr = TRUE)
#> `old[[2]]` is length 1
#> `new[[2]]` is length 0
#> 
#> `old[[2]][[1]]` is an S3 object of class <LayerInstance/Layer/ggproto/gg>, an environment
#> `new[[2]][[1]]` is absent
#> 
#> `old[[3]][[1]]` is length 2
#> `new[[3]][[1]]` is length 0
#> 
#> `old[[3]][[1]][[1]]` is an S3 object of class <ScaleContinuousPosition/ScaleContinuous/Scale/ggproto/gg>, an environment
#> `new[[3]][[1]][[1]]` is absent
#> 
#> `old[[3]][[1]][[2]]` is an S3 object of class <ScaleContinuousPosition/ScaleContinuous/Scale/ggproto/gg>, an environment
#> `new[[3]][[1]][[2]]` is absent
#> 
#> `environment(environment(old[[3]][[2]])[[5]][[1]][[2]])[[5]][[1]]` is length 2
#> `environment(environment(new[[3]][[2]])[[5]][[1]][[2]])[[5]][[1]]` is length 1
#> 
#> `environment(environment(old[[3]][[2]])[[5]][[1]][[2]])[[5]][[1]][[1]]` is a list
#> `environment(environment(new[[3]][[2]])[[5]][[1]][[2]])[[5]][[1]][[1]]` is a function
#> 
#> `environment(environment(old[[3]][[2]])[[5]][[1]][[2]])[[5]][[1]][[2]]` is a function
#> `environment(environment(new[[3]][[2]])[[5]][[1]][[2]])[[5]][[1]][[2]]` is absent
```

**Reply by teunbrand** [reply]
The 3rd case is any time you add anything with `+` via `ggplot2:::add_ggplot()`

**Reply by EvaMaeRey** [reply]
I was more interested in a set_theme_update_to_last() even if side affect-y.  I'm think my problem (classing the list as gg and theme) is a different one than is demonstrated?

``` r
library(ggplot2)
theme_set(theme_dark())
ggplot(cars) +
  aes(x = speed, y = dist) + 
  geom_point() + 
  theme(panel.grid = element_blank())
```

![](https://i.imgur.com/oRa78Pj.png)

``` r

theme_update_last_plot <- function(){
  
  theme_get() %+replace% 
      theme(
      panel.grid = element_blank(),  # would like to use last_plot()$theme somehow here
      complete = TRUE   
      ) 
  
}

theme_set(theme_update_last_plot())

ggplot(mtcars) + 
  aes(wt, mpg) + 
  geom_point()
```

![](https://i.imgur.com/hCiOaB6.png)

<sup>Created on 2024-12-09 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>

**Reply by teunbrand** [reply]
Something like the following?

```r
theme_update_last_plot <- function(plot = last_plot()){
  theme_get() + plot$theme
}
```

**Reply by EvaMaeRey** [reply]
That seems to work, yes!   I didn't expect `theme_get() + last_plot()$theme` to work as an input for theme_set()!  I kind of like it because it's exploratory in theme construction (inspect as you go) and then you and reuse your work w/o copy-paste.  I might not fully appreciate the side affects concerns.

``` r
library(ggplot2)

theme_set_last_plot <- function(){
  
  theme_set(theme_get() + last_plot()$theme)
  
}

ggplot(cars) +
  aes(x = speed, y = dist) + 
  geom_point() +
  theme(panel.grid = element_blank())
```

![](https://i.imgur.com/lmh2tAj.png)

``` r

theme_set_last_plot()

ggplot(mtcars) + 
  aes(wt, mpg) + 
  geom_point()
```

![](https://i.imgur.com/mjXdSKZ.png)

<sup>Created on 2024-12-09 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>

**Reply by EvaMaeRey** [reply]
New application w/ focus on paper, ink, accent focus!   So easy to create ad-hoc chalkboard theme - w/ chalk_themed geoms! 🤯

``` r
library(ggplot2)

# theme_set_last_plot <- function(){
#   
#   theme_set(theme_get() + last_plot()$theme)
#   
# }

ggplot(cars) +
  aes(x = speed, y = dist) + 
  geom_point() +
  theme_classic(ink = alpha("lightyellow", .7),
                paper = "darkseagreen4",
                base_size = 18) + 
  geom_smooth() + 
  theme(geom = element_geom(accent = "orange")) + 
  theme(plot.title.position = "plot") +
  labs(title = "My ad-hoc chalkboard theme,\nwith new ink, paper, accent arguments!") ->
p; p
#> `geom_smooth()` using method = 'loess' and formula = 'y ~ x'
```

![](https://i.imgur.com/OQejLUG.png)

``` r
  
(theme_get() + last_plot()$theme) |> theme_set()

ggplot(mtcars) + 
  aes(wt, mpg) + 
  geom_point() + 
  geom_smooth() + 
  labs(title = "Themed consistent w/ previous plot where\ninteractive theming work was done!")
#> `geom_smooth()` using method = 'loess' and formula = 'y ~ x'
```

![](https://i.imgur.com/z8yF8QJ.png)

<sup>Created on 2024-12-17 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>
